c10/core/TensorImpl.cpp

```diff
@@ -308,6 +315,8 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 AutogradMetaInterface::~AutogradMetaInterface() {}
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
+  TORCH_CHECK(!requires_grad || !is_inference_tensor() || c10::InferenceMode::is_enabled(),
+    "Setting requires_grad=True on inference tensor outside InferenceMode is not allowed.");
```

Comment:
nit: Probably easier to understand prior to demorganization; `!(requires_grad && is_inference_tensor() && !c10::InferenceMode::is_enabled())` because this matches the English text here (I had to mentally reverse the demorgan rule to verify that this condition was correct)

Reply:
It's probably also worth a comment here saying why doing it inside inference mode is OK

Reply:
@ezyang For this one I actually want to keep it this way since in most cases we'd expect this check to pass. And the current way allows early escape from the check if any of the first two conditions is true, which will allow us to only hit `c10::InferenceMode::is_enabled()`(pretty expensive TLS read!) when we have `requires_grad=false` inference tensor case. 
But in the easier to understand version we'll always read TLS. Wdyt?

Reply:
It short circuits in either case. `!requires_grad` in an OR statement short circuits when `!requires_grad` is TRUE; while `requires_grad` in an AND statement short circuits when `requires_grad` is FALSE. If you don't believe me try Godbolt for a change https://godbolt.org/z/YPfh5beoM